### PR TITLE
VZ-5010: Enable Keycloak/Rancher SSO

### PR DESCRIPTION
**Pull request by lake-of-dreams**
- Configures a new client `rancher` in Keycloak with client secret
- Configures a `keycloakoidc` rancher `AuthConfig` CR that sets up the keycloak OIDC provider
- Configures a rancher `User` CR that configures a corresponding user in Rancher
- Configures a rancher `GlobalRoleBinding` CR that assigns admin privs to Verrazzano user
- Entire code in `keycloak` module is refactored to not spin up a `kubectl` process locally in container to execute commands in keycloak pod but rather use the go client. Also refactored all the related tests which earlier spawned bash processes even during testing and had a env variable based way to inject the success/failure condition. Also removed duplicacy from code.
- Also refactored the `k8sutil.ExecPod` utility that will invoke commands in pods, earlier a `fake` for this could only produce one result for any kind of invocation and was problematic when multiple such calls are involved, Changed this to be able to return different `stdout` , `stderr` and `err` based upon the actual command (which eventually becomes the query of the url created by go client) 
- Introduced a func `GetURLForIngress` for getting host url from ingress
-  removed the need of platform-operator/scripts/install/update-kiali-redirect-uris.sh script which was being invoked from keycloak configuration code, instead it was entirely doable using the `-b` parameter on `kcadm` cli
- Changed a test utility func `EventuallyGetRancherURL` to `EventuallyGetURLForIngress` to be able to use at multiple places
- Added logic to disable/enable auth providers

TBD: Logic to disable/enable keycloak vs local authProvider based on setting from rancher component in VZ CR

### CODE REVIEW

File: pkg/k8sutil/k8sutil.go [hunk 1]
```diff
@@ -328,3 +331,13 @@ func GetDynamicClientInCluster(kubeconfigPath string) (dynamic.Interface, error)
...
 }
+
+// GetHostFromIngress returns the url for an Ingress
+func GetURLForIngress(client client.Client, name string, namespace string) (string, error) {
+	var ingress = &networkingv1.Ingress{}
+	err := client.Get(context.TODO(), types.NamespacedName{Name: name, Namespace: namespace}, ingress)
+	if err != nil {
+		return "", fmt.Errorf("unable to fetch ingress %s/%s, %v", name, namespace, err)
+	}
+	return fmt.Sprintf("https://%s", ingress.Spec.Rules[0].Host), nil
```
Comment (kminder) [inline pkg/k8sutil/k8sutil.go:342]: Should the protocol (i.e. https) be passed in for a generic utility like this?

File: platform-operator/controllers/verrazzano/component/keycloak/keycloak.go [hunk 2]
```diff
@@ -285,6 +259,144 @@ const pgClient = `
...
+        "saml.authnstatement": "false",
+        "display.on.consent.screen": "false",
+        "saml.onetimeuse.condition": "false"
+      },
+      "authenticationFlowBindingOverrides": {},
+      "fullScopeAllowed": true,
+      "nodeReRegistrationTimeout": -1,
+      "protocolMappers": [
+        {
+          "id": "8994f0ee-b490-49ae-8793-ee72df7e4b6d",
```
Comment (kminder) [inline platform-operator/controllers/verrazzano/component/keycloak/keycloak.go:310]: Should this and other IDs be generated?

File: platform-operator/controllers/verrazzano/component/keycloak/keycloak.go [hunk 3]
```diff
@@ -1419,3 +1398,149 @@ func upgradeStatefulSet(ctx spi.ComponentContext) error {
...
+
+	err = t.Execute(&b, &data)
+	if err != nil {
+		return "", err
+	}
+
+	return b.String(), nil
+}
+
+// GetRancherClientSecretFromKeycloak returns the secret from rancher client in keycloak
```
Comment (kminder) [inline platform-operator/controllers/verrazzano/component/keycloak/keycloak.go:1429]: Use title case for products in comments.

File: platform-operator/controllers/verrazzano/component/rancher/rancher_component.go [hunk 4]
```diff
@@ -267,6 +267,10 @@ func (r rancherComponent) PostInstall(ctx spi.ComponentContext) error {
 		return log.ErrorfThrottledNewErr("Failed removing Rancher bootstrap secret: %s", err.Error())
 	}
 
+	if err := configureKeycloakOIDCProvider(ctx); err != nil {
+		return log.ErrorfThrottledNewErr("failed configuring keycloak oidc provider: %s", err.Error())
```
Comment (kminder) [inline platform-operator/controllers/verrazzano/component/rancher/rancher_component.go:271]: OIDC?

Comment (lake-of-dreams) [reply platform-operator/controllers/verrazzano/component/rancher/rancher_component.go:271]: its abbreviation for OpenID Connect, the Keycloak OIDC provider `AuthConfig` in Rancher is also called `keycloakoidc`

File: platform-operator/controllers/verrazzano/component/keycloak/keycloak.go (hunk 2 above)
Comment (lake-of-dreams) [reply platform-operator/controllers/verrazzano/component/keycloak/keycloak.go]: removed IDs

File: pkg/k8sutil/k8sutil.go (hunk 1 above)
Comment (lake-of-dreams) [reply pkg/k8sutil/k8sutil.go]: implemented

File: platform-operator/controllers/verrazzano/component/keycloak/keycloak.go (hunk 3 above)
Comment (lake-of-dreams) [reply platform-operator/controllers/verrazzano/component/keycloak/keycloak.go]: fixed

File: platform-operator/helm_config/charts/verrazzano-platform-operator/crds/clusters.verrazzano.io_verrazzanomanagedclusters.yaml [hunk 5]
```diff
@@ -70,7 +70,6 @@ spec:
               rancherRegistration:
                 properties:
                   clusterID:
-                    description: ClusterID is the Rancher cluster ID for this cluster
```
Comment (fredtibbitts) [inline platform-operator/helm_config/charts/verrazzano-platform-operator/crds/clusters.verrazzano.io_verrazzanomanagedclusters.yaml:73]: Why did you remove this description?

Comment (lake-of-dreams) [reply platform-operator/helm_config/charts/verrazzano-platform-operator/crds/clusters.verrazzano.io_verrazzanomanagedclusters.yaml]: excluded the file itself

File: platform-operator/controllers/verrazzano/component/rancher/rancher_component.go [hunk 6]
```diff
@@ -267,6 +269,36 @@ func (r rancherComponent) PostInstall(ctx spi.ComponentContext) error {
...
+		}
+	}
+
+	if vz.Status.Version == "" || (vzStatusVer.IsGreatherThan(ver140) || vzStatusVer.IsEqualTo(ver140)) {
+		enableKeycloak := vz.Spec.Components.Rancher == nil || vz.Spec.Components.Rancher.AuthtType == v1alpha1.Keycloak
+		if err := disableOrEnableAuthProvider(ctx, AuthConfigKeycloak, enableKeycloak); err != nil {
+			return log.ErrorfThrottledNewErr("failed enabling keycloak oidc provider: %s", err.Error())
+		}
+
+		enableLocal := (vz.Spec.Components.Rancher != nil && vz.Spec.Components.Rancher.AuthtType == v1alpha1.Local) || !enableKeycloak
```
Comment (michael-cico) [inline platform-operator/controllers/verrazzano/component/rancher/rancher_component.go:296]: Might be easier on the eye if code like

```
(vz.Spec.Components.Rancher != nil && vz.Spec.Components.Rancher.AuthtType == v1alpha1.Local)
```

was in a helper method like `isLocalAuthEnabled()`, and same for the above KC auth enabled check.

File: platform-operator/controllers/verrazzano/component/rancher/rancher_component.go [hunk 7]
```diff
@@ -302,14 +334,44 @@ func (r rancherComponent) PostUpgrade(ctx spi.ComponentContext) error {
...
+		}
+	}
+
+	if vzSpecVer != nil && (vzSpecVer.IsGreatherThan(ver140) || vzSpecVer.IsEqualTo(ver140)) {
+		enableKeycloak := vz.Spec.Components.Rancher != nil && vz.Spec.Components.Rancher.AuthtType == v1alpha1.Keycloak
+		if err := disableOrEnableAuthProvider(ctx, AuthConfigKeycloak, enableKeycloak); err != nil {
+			return log.ErrorfThrottledNewErr("failed changing state of keycloak oidc provider: %s", err.Error())
+		}
+
+		enableLocal := (vz.Spec.Components.Rancher != nil && vz.Spec.Components.Rancher.AuthtType == v1alpha1.Local) || !enableKeycloak
```
Comment (michael-cico) [inline platform-operator/controllers/verrazzano/component/rancher/rancher_component.go:361]: these have to be mutually exclusive right?  local vs KC auth?  I feel like this can be simplified somehow; i.e.

```
if version >= 1.4.0 {
  if KC-auth {
     // enable local
     // disable KC
     return ...
  } 
}
// enable local auth
// disable KC auth
return ...
```

Also, I think a lot of this code in the PostInstall should be put in a helper-method or something, esp if you need to do something similar in the upgrade path.

Comment (michael-cico) [reply platform-operator/controllers/verrazzano/component/rancher/rancher_component.go:361]: Obviously the above is just some hand-waving, just wondering if the checks are duplicated somehow can be simplified.

**Review by fredtibbitts — COMMENTED** [review]
Some general questions, because I can't tell by looking at the code.

1. Should the default be "local" to preserve the existing behavior?
2. What happens during upgrade? Does the auth provider get changed to Keycloak because that's the default?
3. What if Keycloak is disabled, does the default then become "local"?

File: tests/e2e/verify-infra/restapi/rancher_url_test.go [hunk 8]
```diff
@@ -38,7 +38,7 @@ var _ = t.Describe("rancher", Label("f:infra-lcm",
 				}
 
 				api := pkg.EventuallyGetAPIEndpoint(kubeconfigPath)
-				rancherURL := pkg.EventuallyGetRancherURL(t.Logs, api)
+				rancherURL := pkg.EventuallyGetURLForIngress(t.Logs, api, "cattle-system", "rancher", "https")
```
[the remark this answers is not in the document]
Comment (lake-of-dreams) [reply tests/e2e/verify-infra/restapi/rancher_url_test.go:41]: So currently there is no mechanism for tests to know whether they are running in a upgrade flow or otherwise so I have put some basic tests in place:

1. For version < 1.4.0, will check that local auth config is enabled
2. For version > 1.4.0, will check that one of local or keycloak auth config is enabled because not in all upgrade scenarios will the keycloak be enabled by default

File: platform-operator/controllers/verrazzano/component/rancher/rancher_component.go (hunk 6 above)
Comment (lake-of-dreams) [reply platform-operator/controllers/verrazzano/component/rancher/rancher_component.go]: refactored this

File: platform-operator/controllers/verrazzano/component/rancher/rancher_component.go (hunk 7 above)
Comment (lake-of-dreams) [reply platform-operator/controllers/verrazzano/component/rancher/rancher_component.go]: refactored further please

**Comment by lake-of-dreams** [comment]
@fredtibbitts 
>Should the default be "local" to preserve the existing behavior?
What happens during upgrade? Does the auth provider get changed to Keycloak because that's the default?
What if Keycloak is disabled, does the default then become "local"?

1. As per requirement from Jira, all installs greater than equal to 1.4.0 will have Keycloak as default Auth provider
2. During upgrade there are couple of scenarios
- If from a lower version than 1.4.0 to 1.4.0 or higher, Auth provider will NOT be changed to Keycloak unless the rancher component has `authType` as `kecyloak` in the VZ CR
- if from a version greater than equal to 1.4.0 to a higher version, unless the `authType` is changed in VZ CR while upgrading, it will keep the previous. However if while upgrading the `rancher` section is removed completely, then it will make Keycloak as the auth provider
3. If Keycloak is disabled, only local authProvider be enabled.

File: tests/e2e/verify-infra/restapi/rancher_url_test.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (michael-cico) [reply tests/e2e/verify-infra/restapi/rancher_url_test.go:41]: yeah I think our current pipelines are not flexible enough yet.  The 3 upgrade scenarios are 

1. preserve the local auth
2. user updates to KC auth on upgrade, or post-upgrade
3. User updates to KC with the upgrade

We might be able to achieve both by having the upgrade preserve it in the 1.x -> 1.4+ scenario, and then do a post-upgrade update to KC auth.  But I think that's a separate PR as you say.  The basic upgrade path covers (1) and the post-upgrade update covers (2).  I think (3) is probably a nice-to-have, but requires more flexibility the pipeline and also may not be worth the resources for an extra run.

File: platform-operator/controllers/verrazzano/component/rancher/rancher_component.go [hunk 9]
```diff
@@ -405,5 +353,107 @@ func configureKeycloakOIDCProvider(ctx spi.ComponentContext) error {
...
+		version = vzTargetVersion
+	} else {
+		version = vzSourceVersion
+	}
+
+	return version == nil || (version.IsGreatherThan(ver140) || version.IsEqualTo(ver140)), nil
+}
+
+func isKeycloakAuthEnabled(isUpgrade bool, vz *vzapi.Verrazzano, ver140 *semver.SemVersion, vzSourceVersion *semver.SemVersion, vzTargetVersion *semver.SemVersion) bool {
+	if vz.Spec.Components.Keycloak != nil && vz.Spec.Components.Keycloak.Enabled != nil && !*vz.Spec.Components.Keycloak.Enabled {
```
Comment (michael-cico) [inline platform-operator/controllers/verrazzano/component/rancher/rancher_component.go:401]: there's an `IsKeycloakEnabled()` func in `platform-operator/internal/vzconfig/enabled.go` you could use here

File: platform-operator/controllers/verrazzano/component/rancher/rancher_component.go [hunk 10]
```diff
@@ -405,5 +353,107 @@ func configureKeycloakOIDCProvider(ctx spi.ComponentContext) error {
...
+	if isUpgrade {
+		version = vzTargetVersion
+	} else {
+		version = vzSourceVersion
+	}
+
+	return version == nil || (version.IsGreatherThan(ver140) || version.IsEqualTo(ver140)), nil
+}
+
+func isKeycloakAuthEnabled(isUpgrade bool, vz *vzapi.Verrazzano, ver140 *semver.SemVersion, vzSourceVersion *semver.SemVersion, vzTargetVersion *semver.SemVersion) bool {
```
Comment (michael-cico) [inline platform-operator/controllers/verrazzano/component/rancher/rancher_component.go:400]: Should probably add some comments in this func to make it clear what it's doing, i.e. what the conditions for returning `true` are.

File: platform-operator/controllers/verrazzano/component/rancher/rancher_component.go [hunk 11]
```diff
@@ -405,5 +353,107 @@ func configureKeycloakOIDCProvider(ctx spi.ComponentContext) error {
...
+	if vz.Spec.Components.Rancher != nil && vz.Spec.Components.Rancher.AuthtType == v1alpha1.Keycloak {
+		return true
+	}
+
+	if vz.Spec.Components.Rancher == nil {
+		if !isUpgrade {
+			return true
+		}
+
+		if vzSourceVersion != nil && (vzSourceVersion.IsGreatherThan(ver140) || vzSourceVersion.IsEqualTo(ver140)) && vzTargetVersion != nil && (vzTargetVersion.IsGreatherThan(vzSourceVersion) || vzTargetVersion.IsEqualTo(vzSourceVersion)) {
```
Comment (michael-cico) [inline platform-operator/controllers/verrazzano/component/rancher/rancher_component.go:414]: nit-pick, might be useful to add `IsGreaterThanOrEqualTo` methods on the `SemVer` type.

**Review by wmhopkins — DISMISSED** [review]
lgtm.

File: platform-operator/controllers/verrazzano/component/rancher/rancher_component.go (hunk 9 above)
Comment (lake-of-dreams) [reply platform-operator/controllers/verrazzano/component/rancher/rancher_component.go]: implemented

File: platform-operator/controllers/verrazzano/component/rancher/rancher_component.go (hunk 10 above)
Comment (lake-of-dreams) [reply platform-operator/controllers/verrazzano/component/rancher/rancher_component.go:409]: added comments

File: platform-operator/controllers/verrazzano/component/rancher/rancher_component.go (hunk 11 above)
Comment (lake-of-dreams) [reply platform-operator/controllers/verrazzano/component/rancher/rancher_component.go]: added

**Comment by lake-of-dreams** [comment]
As per feedback from @gilbode , now the `local` auth provider will not be disabled anymore while enabling the `keycloak` auth provider. Changes implemented.

**Review by michael-cico — APPROVED** [review]

File: platform-operator/apis/verrazzano/v1beta1/verrazzano_types.go [hunk 12]
```diff
@@ -608,6 +608,9 @@ type RancherComponent struct {
 	// +optional
 	Enabled          *bool `json:"enabled,omitempty"`
 	InstallOverrides `json:",inline"`
+	// KeycloakAuthEnabled specifies whether the Keycloak Auth provider is enabled.  Default is false.
+	// +optional
+	KeycloakAuthEnabled *bool `json:"keycloakAuthEnabled,omitempty"`
```
Comment (anders-swanson) [inline platform-operator/apis/verrazzano/v1beta1/verrazzano_types.go:613]: FYI if we are adding fields to the API, we need to implement the conversion.
